### PR TITLE
Adding implementation for methods

### PR DESCRIPTION
**Pull request by mdumbrava20**

### CODE REVIEW

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 1]
```diff
@@ -2,68 +2,113 @@
...
+                        .thenComparing(User::getFirstName))
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .distinct()
+                .flatMap(user -> user.getPrivileges().stream())
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: Use method reference ::

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 2]
```diff
@@ -2,68 +2,113 @@
...
+
     }
 
     @Override
     public Map<String, Long> getNumberOfLastNames(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(User::getLastName)
+                .collect(Collectors
+                        .groupingBy(Function.identity(), Collectors.counting()));
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: remove map, put User::getLastName in groupingBy as first parameter

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 3]
```diff
@@ -2,68 +2,113 @@
...
+                        .reduce(x -> true, Predicate::and))
+                .collect(Collectors.toList());
     }
 
     @Override
     public String convertTo(final List<User> users, final String delimiter, final Function<User, String> mapFun) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(mapFun)
+                .collect(Collectors.joining(","));
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: replace "," with delimiter

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 4]
```diff
@@ -2,68 +2,113 @@
...
+                        .groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToInt(User::getAge)
+                .average()
+                .orElse(-1d);
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: Uppercase D

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 5]
```diff
@@ -2,68 +2,113 @@
...
+        Map<String, Long> lastNameFreq = users.stream()
+                .map(User::getLastName)
+                .collect(Collectors.
+                        groupingBy(Function.identity(), Collectors.counting()));
+        Long max = lastNameFreq.values().stream().max(Long::compareTo).orElse(0L);
+        return lastNameFreq.values().stream()
+                .filter(value -> value.equals(max))
+                .count() != 1 ? Optional.empty() : lastNameFreq.entrySet().stream()
+                .max(Map.Entry.comparingByValue()).map(Map.Entry::getKey);
+
```
Comment (vermurachi) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: An alternative solution, could be this:
 return users.stream()
                .collect(groupingBy(User::getLastName, counting()))
                .entrySet().stream().
                        collect(groupingBy(Map.Entry::getValue))
                .entrySet().stream()
                .max(Map.Entry.comparingByKey())
                .map(Map.Entry::getValue)
                .filter(v -> v.size() == 1)
                .map(v -> v.get(0).getKey());

Try to understand how it works and you can use it if you like it more

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 1 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Can't replace with method reference.
img

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 2 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Fixed.

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 3 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Fixed.

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 4 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Fixed.

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 5 above)
Comment (mdumbrava20) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Fixed and understand.

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 1 above)
Comment (vermurachi) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: users.stream().
                map(User::getPrivileges).
                flatMap(Collection::stream).
                distinct().
                collect(Collectors.toList());

Comment (mdumbrava20) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: Fixed.

**Review by vermurachi — APPROVED** [review]

**Pull request merged**